File: FactoryMethod02.js
// SUPERCLASSE CONSTRUCTOR
class Produto{
    constructor(nome, preco){
        this.nome = nome;
        this.preco = preco;
    }

    exibirDetalhes(){
        console.log(`Produto:${this.nome}, Preço: R$${this.preco}`);
    }
}

// SUBCLASSES DE PRODUTOS
class Eletronico extends Produto {
    constructor(nome, preco){
        super(nome, preco);
    }
}

class Roupa extends Produto {
    constructor(nome, preco){
        super(nome, preco);
    }
}

// FABRICA ABSTRATA DE PRODUTOS
class FabricaDeProdutos {
    criarProduto(nome, preco){
        throw  new Error('O método criarProduto deve ser implementado pelas subclasses')
    }
}

// FABRICA CONCRETA - PRODUTO ELETRONICO
class FabricaDeEletronicos extends FabricaDeProdutos{
    criarProduto(nome, preco){
        return new Eletronico(nome, preco);
    }
}

// FABRICA CONCRETA - PRODUTO ROUPAS
class FabricaDeRoupas extends FabricaDeProdutos{
    criarProduto(nome, preco){
        return new Roupa(nome, preco);
    }
}

// INTERFACE DE USO ###########################################################
const fabricaDeEletronicos1 = new FabricaDeEletronicos();
const celular = fabricaDeEletronicos1.criarProduto('Celular', 500);
celular.exibirDetalhes(); 

const fabricaDeRoupas1 = new FabricaDeRoupas();
const camiseta = fabricaDeRoupas1.criarProduto('Camiseta', 50);
camiseta.exibirDetalhes();

const fabricaDeEletronicos2 = new FabricaDeEletronicos();
const notebook = fabricaDeEletronicos1.criarProduto('Notebook', 3500);
notebook.exibirDetalhes(); 

File: FactoryMethod02.py
#SUPERCLASSE CONSTRUCTOR
class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco  = preco

    def exibirDetalhes(self):
        print(f"Nome: {self.nome},Preço: {self.preco},")

# Subclasses de veiculo
class Eletronico(Produto):
    def __init__(self, nome,preco):
        super().__init__(nome,preco)

class Roupa(Produto):
    def __init__(self, nome,preco):
        super().__init__(nome,preco)        

# Fabrica abstrata de veiculo
class FabricaDeProdutos:
    def criarProduto(self, nome,preco):
        raise NotImplementedError("O método criarProduto deve ser implementado pelas subclasses")

# Fabrica concreta de carros
class FabricaDeEletronicos(FabricaDeProdutos):
    def criarProduto(self, nome,preco):
        return Eletronico(nome,preco)

# Fabrica concreta de motos
class FabricaDeRoupas(FabricaDeProdutos):
    def criarProduto(self, nome,preco):
        return Roupa(nome,preco)

# USO DAS FABRICAS
fabricaDeEletronicos1 = FabricaDeEletronicos()
celular = fabricaDeEletronicos1.criarProduto('celular',500)
celular.exibirDetalhes()

fabricaDeEletronicos2 = FabricaDeEletronicos()
notebook = fabricaDeEletronicos2.criarProduto('notebook',3500)
notebook.exibirDetalhes()

fabricaDeRoupas = FabricaDeRoupas()
camiseta = fabricaDeRoupas.criarProduto('camiseta',50)
camiseta.exibirDetalhes()